<launch>

  <arg name="ouster_ns" default="ouster" doc="Override the default namespace of all ouster nodes"/>
  <arg name="sensor_hostname" doc="hostname or IP in dotted decimal form of the sensor"/>
  <arg name="udp_dest" doc="hostname or multicast group IP where the sensor will send UDP data packets"/>
  <arg name="mtp_dest" default=" " doc="hostname IP address for receiving data packets via multicast,
    by default it is INADDR_ANY, so packets will be received on first available interface"/>
  <arg name="mtp_main" default="false" doc="if true, then configure and reinit the sensor, otherwise
    start client with active configuration of sensor"/>
  <arg name="lidar_port" default="0" doc="port to which the sensor should send lidar data"/>
  <arg name="imu_port" default="0" doc="port to which the sensor should send imu data"/>
  <arg name="udp_profile_lidar" default=" " doc="lidar packet profile; possible values: {
    LEGACY,
    RNG19_RFL8_SIG16_NIR16_DUAL,
    RNG19_RFL8_SIG16_NIR16,
    RNG15_RFL8_NIR8
    }"/>
  <arg name="lidar_mode" default=" " doc="resolution and rate; possible values: {
    512x10,
    512x20,
    1024x10,
    1024x20,
    2048x10,
    4096x5
    }"/>
  <arg name="timestamp_mode" default=" " doc="method used to timestamp measurements; possible values: {
    TIME_FROM_INTERNAL_OSC,
    TIME_FROM_SYNC_PULSE_IN,
    TIME_FROM_PTP_1588,
    TIME_FROM_ROS_TIME
    }"/>
  <arg name="metadata" default=" " doc="path to write metadata file when receiving sensor data"/>
  <arg name="viz" default="true" doc="whether to run a rviz"/>
  <arg name="rviz_config" default="$(find ouster_ros)/config/viz.rviz" doc="optional rviz config file"/>

  <arg name="tf_prefix" default=" " doc="namespace for tf transforms"/>
  <arg name="sensor_frame" default="os_sensor"
    doc="sets name of choice for the sensor_frame tf frame, value can not be empty"/>
  <arg name="lidar_frame" default="os_lidar"
    doc="sets name of choice for the os_lidar tf frame, value can not be empty"/>
  <arg name="imu_frame" default="os_imu"
    doc="sets name of choice for the os_imu tf frame, value can not be empty"/>
  <arg name="point_cloud_frame" default=" "
    doc="which frame to be used when publishing PointCloud2 or LaserScan messages.
    Choose between the value of sensor_frame or lidar_frame, leaving this value empty
    would set lidar_frame to be the frame used when publishing these messages."/>

  <arg name="no_bond" default="false"
    doc="request no bond setup when nodelets are created"/>
  <arg if="$(arg no_bond)" name="_no_bond" value="--no-bond"/>
  <arg unless="$(arg no_bond)" name="_no_bond" value=" "/>

  <arg name="dynamic_transforms_broadcast" default="false"
    doc="static or dynamic transforms broadcast"/>
  <arg name="dynamic_transforms_broadcast_rate" default="1.0"
    doc="set the rate (Hz) of broadcast when using dynamic broadcast; minimum value is 1 Hz"/>

  <arg name="proc_mask" default="IMG|PCL|IMU|SCAN" doc="
    The IMG flag here is not supported and does not affect anything,
    to disable image topics you would need to omit the os_image node
    from the launch file"/>

  <arg name="scan_ring" default="0" doc="
    use this parameter in conjunction with the SCAN flag
    and choose a value the range [0, sensor_beams_count)"/>

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet" name="os_nodelet_mgr"
      output="screen" required="true"
      launch-prefix="bash -c 'sleep 2; $0 $@' "
      args="manager"/>
  </group>

  <group ns="$(arg ouster_ns)">
    <node pkg="nodelet" type="nodelet" name="os_node"
      output="screen" required="true"
      args="load ouster_ros/OusterSensor os_nodelet_mgr $(arg _no_bond)">
      <param name="~/sensor_hostname" type="str" value="$(arg sensor_hostname)"/>
      <param name="~/udp_dest" type="str" value="$(arg udp_dest)"/>
      <param name="~/mtp_dest" type="str" value="$(arg mtp_dest)"/>
      <param name="~/mtp_main" type="bool" value="$(arg mtp_main)"/>
      <param name="~/lidar_port" type="int" value="$(arg lidar_port)"/>
      <param name="~/imu_port" type="int" value="$(arg imu_port)"/>
      <param name="~/udp_profile_lidar" type="str" value="$(arg udp_profile_lidar)"/>
      <param name="~/lidar_mode" type="str" value="$(arg lidar_mode)"/>
      <param name="~/timestamp_mode" type="str" value="$(arg timestamp_mode)"/>
      <param name="~/metadata" type="str" value="$(arg metadata)"/>
    </node>
  </group>

  <include file="$(find ouster_ros)/launch/common.launch">
    <arg name="ouster_ns" value="$(arg ouster_ns)"/>
    <arg name="viz" value="$(arg viz)"/>
    <arg name="rviz_config" value="$(arg rviz_config)"/>
    <arg name="tf_prefix" value="$(arg tf_prefix)"/>
    <arg name="sensor_frame" value="$(arg sensor_frame)"/>
    <arg name="lidar_frame" value="$(arg lidar_frame)"/>
    <arg name="imu_frame" value="$(arg imu_frame)"/>
    <arg name="point_cloud_frame" value="$(arg point_cloud_frame)"/>
    <arg name="timestamp_mode" value="$(arg timestamp_mode)"/>
    <arg name="_no_bond" value="$(arg _no_bond)"/>
    <arg name="dynamic_transforms_broadcast"
      value="$(arg dynamic_transforms_broadcast)"/>
    <arg name="dynamic_transforms_broadcast_rate"
      value="$(arg dynamic_transforms_broadcast_rate)"/>
    <arg name="proc_mask" value="$(arg proc_mask)"/>
    <arg name="scan_ring" value="$(arg scan_ring)"/>
  </include>

</launch>
